src/main/java/sparksoniq/semantics/visitor/StaticContextVisitor.java

```diff
@@ -48,76 +49,109 @@
 
     @Override
     public StaticContext visit(ExpressionOrClause expression, StaticContext argument) {
-        if (expression instanceof Expression)
+        if (expression instanceof Expression) {
             ((Expression) expression).setStaticContext(argument);
+        }
         return expression.accept(this, argument);
     }
 
+    // region primary
     @Override
     public StaticContext visitVariableReference(VariableReference expression, StaticContext argument) {
-        if (argument == null || !argument.isInScope(expression.getVariableName()))
+        String variableName = expression.getVariableName();
+        if (argument == null || !argument.isInScope(variableName)) {
             throw new UndeclaredVariableException(
-                    "Uninitialized variable reference: " + expression.getVariableName(),
+                    "Uninitialized variable reference: " + variableName,
                     expression.getMetadata()
             );
-        else {
-            expression.setType(argument.getVariableSequenceType(expression.getVariableName()));
+        } else {
+            expression.setType(argument.getVariableSequenceType(variableName));
+            expression.setHighestExecutionMode(argument.getVariableExecutionMode(variableName));
             return defaultAction(expression, argument);
         }
     }
 
+    @Override
+    public StaticContext visitFunctionDeclaration(FunctionDeclaration expression, StaticContext argument) {
+        expression.registerUserDefinedFunctionExecutionMode();
+
+        // define a static context for the function body, add params to the context and visit the body expression
+        StaticContext functionDeclarationContext = new StaticContext(argument);
+        expression.get_params()
+            .forEach(
+                (paramName, flworVarSequenceType) -> functionDeclarationContext.addVariable(
+                    paramName,
+                    flworVarSequenceType.getSequence(),
+                    expression.getMetadata(),
+                    ExecutionMode.LOCAL // static udf currently supports materialized(local) params, not RDDs or DFs
+                )
```

Comment:
At the moment, static function declarations allow parametrized variables to be in the materialized form. In other words, (and if I'm not mistaken), user defined functions would not work properly when a variable that is a DF or an RDD is passed as a parameter.

Reply:
I agree with always passing materialized parameters to UDFs. This makes the higherExecutionMode of the expressions in the function body not depend on the parameters, and "truly static": indeed, if the same function is called at multiple places, there is no guarantee the passed parameters will always be local or always be RDDs.

If we relax this later, this will require significantly more work to get right; now would be too premature.